akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala

```diff
@@ -81,63 +83,61 @@ import scala.util.{ Failure, Success, Try }
+            }
           }
-        }
-        subSink.setHandler {
-          new InHandler {
-            override def onPush(): Unit = push(out, subSink.grab())
+          subSink.setHandler {
+            new InHandler {
+              override def onPush(): Unit = push(out, subSink.grab())
 
-            override def onUpstreamFinish(): Unit = complete(out)
+              override def onUpstreamFinish(): Unit = complete(out)
 
-            override def onUpstreamFailure(ex: Throwable): Unit = fail(out, ex)
-          }
-        }
-        Try {
-          Source.fromGraph(subSource.source).viaMat(flow)(Keep.right).to(subSink.sink).run()(subFusingMaterializer)
-        } match {
-          case Success(matVal) =>
-            innerMatValue.success(matVal)
-            Initializing.upstreamFailure match {
-              case OptionVal.Some(ex) =>
-                subSource.fail(ex)
-              case OptionVal.None =>
-                if (isClosed(in))
-                  subSource.complete()
-            }
-            Initializing.downstreamCause match {
-              case OptionVal.Some(cause) =>
-                subSink.cancel(cause)
-              case OptionVal.None =>
-                //todo: should this be invoked before and independently of checking downstreamCause?
-                // in most case if downstream pulls and then closes, the pull is 'lost'. is it possible for some flows to actually care about this? (non-eager broadcast?)
-                if (isAvailable(out)) {
-                  subSink.pull()
-                }
+              override def onUpstreamFailure(ex: Throwable): Unit = fail(out, ex)
             }
-          case Failure(ex) =>
-            innerMatValue.failure(new NeverMaterializedException(ex))
-            failStage(ex)
-        }
-        setHandlers(in, out, new InHandler with OutHandler {
-          override def onPull(): Unit = subSink.pull()
-          override def onDownstreamFinish(cause: Throwable): Unit = subSink.cancel(cause)
+          }
+          Try {
+            Source.fromGraph(subSource.source).viaMat(flow)(Keep.right).to(subSink.sink).run()(subFusingMaterializer)
+          } match {
+            case Success(matVal) =>
+              innerMatValue.success(matVal)
+              upstreamFailure match {
+                case OptionVal.Some(ex) =>
+                  subSource.fail(ex)
+                case OptionVal.None =>
+                  if (isClosed(in))
+                    subSource.complete()
+              }
+              downstreamCause match {
+                case OptionVal.Some(cause) =>
+                  subSink.cancel(cause)
+                case OptionVal.None =>
+                  if (isAvailable(out)) subSink.pull()
+              }
+            case Failure(ex) =>
+              innerMatValue.failure(new NeverMaterializedException(ex))
+              failStage(ex)
+          }
+          setHandlers(in, out, new InHandler with OutHandler {
```

Comment:
It's a bit weird that this is done even if the materialization has failed. Maybe replace the `Try {}` with a regular `try`/`catch` block.